ipa-core/src/report/hybrid.rs

```diff
@@ -150,37 +151,94 @@ impl TryFrom<Bytes> for EncryptedHybridReport {
     }
 }
 
+const TAG_SIZE: usize = TagSize::USIZE;
+
+#[derive(Clone, Debug)]
+pub struct UniqueTag {
+    bytes: [u8; TAG_SIZE],
+}
+
 pub trait UniqueBytes {
-    fn unique_bytes(&self) -> Vec<u8>;
+    fn unique_bytes(&self) -> [u8; TAG_SIZE];
+}
+
+impl UniqueBytes for UniqueTag {
+    fn unique_bytes(&self) -> [u8; TAG_SIZE] {
+        self.bytes
+    }
 }
 
 impl UniqueBytes for EncryptedHybridReport {
     /// We use the `TagSize` (the first 16 bytes of the ciphertext) for collision-detection
     /// See [analysis here for uniqueness](https://eprint.iacr.org/2019/624)
-    fn unique_bytes(&self) -> Vec<u8> {
-        self.mk_ciphertext()[0..TagSize::USIZE].to_vec()
+    fn unique_bytes(&self) -> [u8; TAG_SIZE] {
+        let slice = &self.mk_ciphertext()[0..TAG_SIZE];
+        let mut array = [0u8; TAG_SIZE];
+        array.copy_from_slice(slice);
+        array
+    }
+}
+
+impl UniqueTag {
+    // Function to attempt to create a UniqueTag from a UniqueBytes implementor
+    pub fn from_unique_bytes<T: UniqueBytes>(item: &T) -> Self {
+        UniqueTag {
+            bytes: item.unique_bytes(),
+        }
+    }
+
+    /// Maps the tag into a consistent shard.
+    /// Note that `ShardIndex` is limited to u32, so we only use the first 4 bytes.
```

Comment:
Doesn't it introduce bias? Can we lift shard index to u128, do the modulo operation and then downcast to u32?

Reply:
we certainly can, but does it introduce bias? I suppose I'm assuming that all the bytes in the tag are uniformly distributed, so taking 4 random bytes will still be uniformly distributed.

Reply:
I think we can say that about AES tag, but there is no such guarantee on a subset of AAD bytes afaik

Reply:
I'm pretty sure a uniformly distributed tag implies any subsets of tags are also uniformly distributed. (look at it from the other direction, a non-uniformly random bit must result in the full tag being non-uniformly random.)

doesn't matter though, I'll push a change that does it on u128s.